### PR TITLE
Implement vectorized adstock transformations

### LINKED ISSUE #22: Adstock transformation without `for` loop
We would like to write a more efficient implementation of the adstock transformations so that we do not use a for loop. An attempt with `scan` was (unsuccessfully) implemented in https://github.com/pymc-labs/pymmmc/pull/15 

**Requirements:**
- We should be able to add the `l_max` parameter to truncate the size of the effect. 
- Be vectorised
- Should be faster than the current implementation.

### PR DESCRIPTION
**Pull request by lucianopaz**
Closes #22 

- [X] Convert the old adstock transformations to use the vectorized convolution
- [x] Write tests for the vectorized convolution with multiple dimensions and broadcasting patterns
- [x] Maybe add an assert to prevent segfaults due to passing a incorrect axis?

### CODE REVIEW

File: pymmmc/transformers.py [hunk 1]
```diff
@@ -31,29 +82,28 @@ def geometric_adstock(x, alpha: float = 0.0, l_max: int = 12, normalize: bool =
...
-        at.concatenate(tensor_list=[at.zeros(shape=x.shape)[:i], x[: x.shape[0] - i]])
-        for i in range(l_max)
-    ]
-    x_cycle = at.stack(cycles)
-    x_cycle = at.transpose(x=x_cycle, axes=[1, 2, 0])
-    w = at.as_tensor_variable([at.power(alpha, i) for i in range(l_max)])
-    w = at.transpose(w)[None, ...]
-    w = w / at.sum(w, axis=2, keepdims=True) if normalize else w
-    return at.sum(at.mul(x_cycle, w), axis=2)
+    warnings.warn(
```
Comment (twiecki) [inline pymmmc/transformers.py:92]: Just delete it, we haven't released yet.

File: pymmmc/transformers.py [hunk 2]
```diff
@@ -1,7 +1,58 @@
...
+        The result of convolving ``x`` with ``w`` along the desired axis. The shape of the
+        result will match the shape of ``x`` up to broadcasting with ``w``. The convolved
+        axis will show the results of left padding zeros to ``x`` while applying the
+        convolutions.
+    """
+    orig_ndim = x.ndim
+    axis = axis if axis >= 0 else orig_ndim + axis
+    w = at.as_tensor(w)
+    x = at.moveaxis(x, axis, -1)
+    x_shape, w_shape = at.random.utils.params_broadcast_shapes(
```
Comment (lucianopaz) [inline pymmmc/transformers.py:31]: @ricardoV94, is there another way to implement this shape broadcasting?

- If I’m not explicit about it, broadcasting of symbolic `x` and `w` fails and takes the shape of `x` as the correct one (maybe symbolic isn’t the correct term, I mean tensors that have `None` in their shape).
- When I am explicit but the arrays fail to broadcast at run time, I get segfaults (probably due to the C backend trying to read or write to memory locations that aren’t part of the array)

**Comment by drbenvincent** [comment]
Could I recommend:
a) placing an assert at the start of the function to ensure sizes of the data and parameters are aligned
b) is it worth doing some timing tests, even if informal. I'm finding that inference is slowing down significantly when going from [4, T] sized inputs (ie 4 time series) to [4, 150, T] (ie 4 time series each of which has 150 geographical regions.

File: pymmmc/transformers.py (hunk 2 above)
Comment (lucianopaz) [reply pymmmc/transformers.py:31]: From a closer look at [the source](https://github.com/pymc-devs/pytensor/blob/main/pytensor/tensor/random/utils.py#L55) I think that there should be a couple of assert Ops there.

**Comment by drbenvincent** [comment]
Feature request: At the moment there's the assumption that if our data is shape `[4, 150, T]`, then we have parameters of shape `[4, 150]`. This makes sense in the case that we want independent parameters for each predictor*region combination, whether we place a hierarchy on that or not. But, this is likely to significantly slow down inference. 

Would it be possible to have the option (or an alternative function) which allows you to just have different priors over predictors which is applied over all regions. So in this example, that would be data of size `[4, 150, T]`, but parameters of size `[4]`

**Comment by lucianopaz** [comment]
> Feature request: At the moment there's the assumption that if our data is shape `[4, 150, T]`, then we have parameters of shape `[4, 150]`. This makes sense in the case that we want independent parameters for each predictor*region combination, whether we place a hierarchy on that or not. But, this is likely to significantly slow down inference.
> 
> Would it be possible to have the option (or an alternative function) which allows you to just have different priors over predictors which is applied over all regions. So in this example, that would be data of size `[4, 150, T]`, but parameters of size `[4]`

This is already supported, just pass the parameters with shape `(4, 1)` and the same parameter value will be used for every entry in the second axis of `x`. The logic behind this is that the T dimension of `x` ids ignored, and the rest of the axis are broadcasted following the same rules that numpy does. That’s why you got a failure to broadcast when the parameter had shape `(4,)`, because 4 and 150 don’t broadcast together

File: pymc_marketing/mmm/transformers.py [hunk 3]
```diff
@@ -31,29 +82,29 @@ def geometric_adstock(x, alpha: float = 0.0, l_max: int = 12, normalize: bool =
...
-        pt.concatenate(tensor_list=[pt.zeros(shape=x.shape)[:i], x[: x.shape[0] - i]])
-        for i in range(l_max)
-    ]
-    x_cycle = pt.stack(cycles)
-    x_cycle = pt.transpose(x=x_cycle, axes=[1, 2, 0])
-    w = pt.as_tensor_variable([pt.power(alpha, i) for i in range(l_max)])
-    w = pt.transpose(w)[None, ...]
-    w = w / pt.sum(w, axis=2, keepdims=True) if normalize else w
-    return pt.sum(pt.mul(x_cycle, w), axis=2)
+    warnings.warn(
```
Comment (twiecki) [inline pymc_marketing/mmm/transformers.py:93]: lets just remove this, were still pre-release

Comment (lucianopaz) [reply pymc_marketing/mmm/transformers.py]: Done

**Comment by lucianopaz** [comment]
@ricardoV94, I'd like to ask for your input on the `params_broadcast_shapes` function that I added here. As I mentioned in some older comments, the `pytensor` version is not robust to broadcasting failures (shapes that cannot broadcast together are taken to broadcast just fine and can cause segfaults down the line). This version might be suitable for `pytensor` itself, if you think that it's OK. My main question is whether the approach that I followed for getting "concrete shapes" (shapes that are not `TensorVariable`s themselves) is fine, or if there is something that I might have missed.

**Comment by lucianopaz** [comment]
I opened pymc-devs/pytensor#152, related to the problem I found with `params_broadcast_shapes`. If the fix from this PR looks ok, I can also apply it over on `pytensor`

**Comment by lucianopaz** [comment]
Pinging @ricardoV94 to have a look. I updated the `params_shape_broadcast` to rely on the tensors `broadcastable` attribute. If this isn't desirable, I can undo the last commit.

**Comment by ricardoV94** [comment]
You can't set the broadcastable flag manually for the time being (it's a read only property, except for the Unbroadcast Op which was kept just for some edge cases) so I am not sure it's worth copying the method here. When we reintroduce static broadcasting the Pytensor method will be adapted accordingly.

**Comment by juanitorduz** [comment]
Hey! Any blockers for this one? Anything I could do to support ?

**Comment by ricardoV94** [comment]
Would be good to see if we can remove the custom `params_broadcast_shapes` and use the one in PyTensor, now that we fixed `BroadcastTo`

**Comment by ricardoV94** [comment]
Closed in favor of #221

**Pull request closed without merging**